typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts

```diff
@@ -168,4 +168,74 @@ export class HyperlaneCoreChecker<
       this.addViolation(violation);
     }
   }
+
+  async checkInterchainGasPaymaster(local: Chain): Promise<void> {
+    const coreContracts = this.app.getContracts(local);
+    const igp = coreContracts.interchainGasPaymaster.contract;
+
+    // Construct the violation, updating the actual & expected
+    // objects as violations are found.
+    // A single violation is used so that only a single `setGasOracles`
+    // call is generated to set multiple gas oracles.
+    const gasOraclesViolation: IgpGasOraclesViolation = {
+      type: CoreViolationType.InterchainGasPaymaster,
+      subType: IgpViolationType.GasOracles,
+      contract: igp,
+      chain: local,
+      actual: {},
+      expected: {},
+    };
+
+    // The `gasOracles` mapping was added in a new implementation of the IGP.
+    // If calling this reverts, we are still using an old implementation that
+    // must be upgraded, so we just consider the mapping as unset and default
+    // to address(0) for all gas oracles.
+    // Note that it's important that `checkProxiedContracts` is called before
+    // `checkInterchainGasPaymaster` - this way, the govern script will first
+    // change the IGP proxy to use the correct new implementation, and the
+    // IgpGasOraclesViolation will be able to be correctly handled.
+    const getCurrentGasOracle = async (
+      remoteId: number,
+    ): Promise<types.Address> => {
+      try {
+        return await igp.gasOracles(remoteId);
+      } catch (_) {
+        return ethers.constants.AddressZero;
+      }
+    };
+
+    const remotes = this.multiProvider.remoteChains(local);
+    for (const remote of remotes) {
+      const remoteId = ChainNameToDomainId[remote];
+      const actualGasOracle = await getCurrentGasOracle(remoteId);
+      const expectedGasOracle = this.getGasOracleAddress(local, remote);
+
+      if (!utils.eqAddress(actualGasOracle, expectedGasOracle)) {
+        const remoteChain = remote as ChainName;
+        gasOraclesViolation.actual[remoteChain] = actualGasOracle;
+        gasOraclesViolation.expected[remoteChain] = expectedGasOracle;
+      }
+    }
+    // Add the violation only if it's been populated with gas oracle inconsistencies
+    if (Object.keys(gasOraclesViolation.actual).length > 0) {
+      this.addViolation(gasOraclesViolation);
+    }
+  }
+
+  getGasOracleAddress(local: Chain, remote: Chain): types.Address {
+    const config = this.configMap[local];
+    const gasOracleType = config.igp.gasOracles[remote];
+    if (!gasOracleType) {
+      throw Error(
+        `Expected gas oracle type for local ${local} and remote ${remote}`,
+      );
+    }
+    const coreContracts = this.app.getContracts(local);
+    switch (gasOracleType) {
+      case GasOracleContractType.StorageGasOracle:
+        return coreContracts.storageGasOracle.address;
+      default:
+        throw Error(`Unsupported gas oracle type ${gasOracleType}`);
+    }
```

Comment:
imo we should just have a `coreContracts.gasOracle` and the type should be on config but transparent to core

Reply:
actually.. is there an easy way of doing this?

I think this would require splitting `CoreConfig` into 2 types or something - one where we can configure the `GasOracleContractType` and another where it uses the address of the gas oracle?
Not really sure how that can be done without tons of changes, can you point me in the right direction?

Reply:
Would like to avoid needing to specify the addresses in `infra/config/environments/<env>/core.ts` because then deploying would need to be a two step thing